packages/browser/src/profiling/lifecycleMode/traceLifecycleProfiler.ts

```diff
@@ -0,0 +1,312 @@
+    this._client = undefined;
+    this._profiler = undefined;
+    this._chunkTimer = undefined;
+    this._activeRootSpanCount = 0;
+    this._activeRootSpanIds = new Set<string>();
+    this._profilerId = undefined;
+    this._isRunning = false;
+    this._sessionSampled = false;
+  }
+
+  /**
+   * Initialize the profiler with client and session sampling decision computed by the integration.
+   */
+  public initialize(client: Client, sessionSampled: boolean): void {
+    DEBUG_BUILD && debug.log("[Profiling] Initializing profiler (lifecycle='trace').");
+
+    this._client = client;
+    this._sessionSampled = sessionSampled;
+
+    client.on('spanStart', span => {
+      if (span !== getRootSpan(span)) {
+        return;
+      }
+      if (!this._sessionSampled) {
+        DEBUG_BUILD && debug.log('[Profiling] Session not sampled because of negative sampling decision.');
+        return;
+      }
+      // Only count sampled root spans
+      if (!span.isRecording()) {
+        DEBUG_BUILD && debug.log('[Profiling] Discarding profile because root span was not sampled.');
+        return;
+      }
+
+      const rootSpanJSON = spanToJSON(span);
+      const spanId = rootSpanJSON.span_id as string | undefined;
+      if (!spanId) {
+        return;
+      }
+      if (this._activeRootSpanIds.has(spanId)) {
+        return;
+      }
+      this._activeRootSpanIds.add(spanId);
+
+      const wasZero = this._activeRootSpanCount === 0;
+      this._activeRootSpanCount++; // Increment before eventually starting the profiler
+      DEBUG_BUILD &&
+        debug.log(
+          `[Profiling] Root span ${rootSpanJSON.description} started. Active root spans:`,
+          this._activeRootSpanCount,
+        );
+      if (wasZero) {
+        this.start();
+      }
+    });
+
+    client.on('spanEnd', span => {
+      if (!this._sessionSampled) {
+        return;
+      }
+
+      const spanJSON = spanToJSON(span);
+      const spanId = spanJSON.span_id as string | undefined;
+      if (!spanId || !this._activeRootSpanIds.has(spanId)) {
+        return;
+      }
+
+      this._activeRootSpanIds.delete(spanId);
+      this._activeRootSpanCount = Math.max(0, this._activeRootSpanCount - 1);
+      DEBUG_BUILD &&
+        debug.log(
+          `[Profiling] Root span ${spanJSON.description} ended. Active root spans: ${this._activeRootSpanCount}`,
+        );
+      if (this._activeRootSpanCount === 0) {
+        this._collectCurrentChunk().catch(() => /* no catch */ {});
+
+        this.stop();
+      }
+    });
+  }
+
+  /**
+   * Handle an already-active root span at integration setup time.
+   */
+  public notifyRootSpanActive(span: Span): void {
+    if (!this._sessionSampled) {
+      return;
+    }
+    if (span !== getRootSpan(span)) {
+      return;
+    }
+    const spanId = spanToJSON(span)?.span_id;
+    if (!spanId || this._activeRootSpanIds.has(spanId)) {
+      return;
+    }
+
+    const wasZero = this._activeRootSpanCount === 0;
```

Comment:
Same comment as above on simplifying the logic

Reply:
I deleted `_activeRootSpanCount` now in general as I can just use `this._activeRootSpanIds.size`.